modules/core/04-channel/types/multihop.go

```diff
@@ -1,45 +1,53 @@
 package types
 
 import (
-	"fmt"
-
+	errorsmod "cosmossdk.io/errors"
 	"github.com/cosmos/cosmos-sdk/codec"
 	connectiontypes "github.com/cosmos/ibc-go/v8/modules/core/03-connection/types"
+	ibcerrors "github.com/cosmos/ibc-go/v8/modules/core/errors"
 	"github.com/cosmos/ibc-go/v8/modules/core/exported"
 )
 
 type ConnectionEnd = connectiontypes.ConnectionEnd
 
 // TODO: Create multihop proof struct to serialize multihop proofs into.
 
-// GetMultihopConnectionEnd returns the final connectionEnd from the counterparty perspective
-func (m *MsgMultihopProofs) GetMultihopConnectionEnd(cdc codec.BinaryCodec, connection exported.ConnectionI) (*ConnectionEnd, error) {
+// GetLastHopConnectionEnd returns the last hop connectionEnd from the perpespective of the executing chain.
+// The last hop connection is the connection end on the chain before the counterparty multihop chain (i.e. the chain on the other end of the multihop channel).
+func (m *MsgMultihopProofs) GetLastHopConnectionEnd(cdc codec.BinaryCodec, connection exported.ConnectionI) (*ConnectionEnd, error) {
 	var connectionEnd ConnectionEnd
 	if len(m.ConnectionProofs) > 0 {
+		// connection proofs are ordered from executing chain to counterparty,
+		// so the last hop connection end is the value of last connection proof
 		if err := cdc.Unmarshal(m.ConnectionProofs[len(m.ConnectionProofs)-1].Value, &connectionEnd); err != nil {
 			return nil, err
 		}
 	} else {
```

Comment:
Yeah, it's been a bit since I wrote this, but I believe I designed the multi-hop code so that it could also work for the case of a single-hop. In that case, no extra connection proofs are required because we already store the IBC connection for the counterparty. i.e. this is handling the degenerate case when multi-hop works like single-hop.

I was hoping that somehow the single-hop and multi-hop verification code could merge one day.

Reply:
>I was hoping that somehow the single-hop and multi-hop verification code could merge one day.

Yes, that would be great! I was also thinking about it the other day. It would make the code easier to reason about. I will try to see if I can come up with something, or ask later on the rest of the team if they would have any ideas.

Reply:
Currently the connectionKeeper interface has verification methods for each separate proof type (channel try/ack, recv packer, ack packet, etc). But really they are only needed because the derived key/values are passed in and need to be computed differently for each case. The multihop verification method allows passing in a key/value generation function which could generalize this.

The next issue is the proof format being different, but if verification could work similarly given a single-hop or multi-hop proof method its seems possible to collapse all the methods down to: VerifyMembership and VerifyNonmembership.

Anyway, these were a few thoughts I had a while back.